src/utilities/createRouter.ts

```diff
@@ -90,21 +94,65 @@ export function createRouter<const T extends RouterRoutes>(routes: T, options: R
     }
   }
 
-  const push: RouterPush<T> = (source, options) => {
-    const url = resolve(source, options)
-
-    return update(url, { replace: options?.replace })
+  function push<TSource extends RouteKeysThatHaveRequireParams<T>>(
+    source: TSource,
+    params: RouteParamsByName<T, TSource>,
+    options?: RouterPushOptions
+  ): Promise<void>
+  function push<TSource extends RoutesKey<T>>(
+    source: TSource,
+    params?: RouteParamsByName<T, TSource>,
+    options?: RouterPushOptions
+  ): Promise<void>
+  function push(
+    source: Url,
+    options?: RouterPushOptions
+  ): Promise<void>
+  function push(
+    source: any,
+    paramsOrOptions?: any,
+    maybeOptions?: RouterPushOptions,
+  ): Promise<void> {
+    const options: RouterPushOptions = (isUrl(source) ? paramsOrOptions : maybeOptions) ?? {}
+    const url = resolve(source, paramsOrOptions, maybeOptions)
+
+    return update(url, { replace: options.replace })
   }
 
-  const replace: RouterReplace<T> = (source, options) => {
-    return push(source, { ...options, replace: true })
+  function replace<TSource extends RouteKeysThatHaveRequireParams<T>>(
+    source: TSource,
+    params: RouteParamsByName<T, TSource>,
+    options?: RouterReplaceOptions
+  ): Promise<void>
+  function replace<TSource extends RoutesKey<T>>(
+    source: TSource,
+    params?: RouteParamsByName<T, TSource>,
+    options?: RouterReplaceOptions
+  ): Promise<void>
+  function replace(
+    source: Url,
+    options?: RouterReplaceOptions
+  ): Promise<void>
+  function replace<TSource extends Url | RoutesKey<T>>(
+    source: TSource,
+    paramsOrOptions?: Record<string, unknown> | RouterPushOptions,
+    maybeOptions?: RouterReplaceOptions,
+  ): Promise<void> {
+    if (isUrl(source)) {
+      const options: RouterPushOptions = paramsOrOptions ?? {}
+      return push(source, { ...options, replace: true })
+    }
+
+    const params: any = paramsOrOptions
+    const options: RouterPushOptions = maybeOptions ?? {}
+    return push(source, params, { ...options, replace: true })
```

Comment:
Replace could be defined the same way. Here you have more specific implementation types, but if we're consistent with `push` it seems to work from my testing
```
   const replace: RouterReplace<T> = (source: any, paramsOrOptions?: any, maybeOptions?: RouterReplaceOptions): Promise<void> => {
    if (isUrl(source)) {
      const options = paramsOrOptions ?? {}
      return push(source, { ...options, replace: true })
    }

    const params = paramsOrOptions
    const options = maybeOptions ?? {}

    return push(source, params, { ...options, replace: true })
  }
```